[homematic] Invalid ID for hue integration causes errors

I am using the new (beta) hue integration in my CCU2.
After pairing the hue bridge and the CCU2 the hue bridge (and the associated devices) show up in the CCU device list.
Unfortunately the hue bridge is added in the CCU with the following ID: "HU-Philips hue"
The ID can not be changed in the CCU, and contains a whitespace.


## Expected Behavior
Everything keeps working, even with a hue bridge paired with the CCU2.
Hue devices can be used in openhab.

## Current Behavior
The binding doesn't accept the hue bridge as valid device and shows the following error:
`[ERROR] [very.HomematicDeviceDiscoveryService] - UID segment 'HU-Philips hue' contains invalid characters. Each segment of the UID must match the pattern [A-Za-z0-9_-]*.`

Unfortunately this error seems to mess up some more things, as the Gateway-Extras are not working any longer.
Log:
`[hingStatusInfoChangedEvent] - 'homematic:GATEWAY-EXTRAS-CCU:ccu:GWE00000000' changed from OFFLINE (CONFIGURATION_ERROR): Device with address 'GWE00000000' not found on gateway 'ccu' to UNINITIALIZED`

After deleting the hue devices in the CCU everything is working again in openhab without any configuration changes in openhab!
So the hue integration in the CCU seems to be the root problem. (probably the whitespace in the ID)

## Possible Solution
Replace the whitespace in the device id by "_" as it is done for variables.

## Steps to Reproduce (for Bugs)
1. Pair a hue in the CCU. 
2. Try to use GATEWAY-EXTRAS in openhab.

## Context
-

## Your Environment
CCU2 Firmware:2.31.25
XML-Api: 1.15
Cux-D: 1.12
Openhab: 2.2 (short test on 2.3 showed same behaviour)

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/CcuLoadDeviceNamesParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/CcuLoadDeviceNamesParser.java
@@ -38,7 +38,7 @@ public class CcuLoadDeviceNamesParser extends CommonRpcParser<TclScriptDataList,
             }
 
             for (TclScriptDataEntry entry : resultList.getEntries()) {
-                HmDevice device = devicesByAddress.get(getAddress(entry.name));
+                HmDevice device = devicesByAddress.get(getSanitizedAddress(entry.name));
                 if (device != null) {
                     device.setName(entry.value);
                 }

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/CommonRpcParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/CommonRpcParser.java
@@ -108,9 +108,9 @@ public abstract class CommonRpcParser<M, R> implements RpcParser<M, R> {
     }
 
     /**
-     * Returns the address of a device, replacing group address identifier.
+     * Returns the address of a device, replacing group address identifier and illegal characters.
      */
-    protected String getAddress(Object object) {
+    protected String getSanitizedAddress(Object object) {
         String address = StringUtils.trimToNull(StringUtils.replaceOnce(toString(object), "*", "T-"));
         return MiscUtils.validateCharacters(address, "Address", "_");
     }

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/CommonRpcParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/CommonRpcParser.java
@@ -14,6 +14,7 @@ import org.apache.commons.lang.BooleanUtils;
 import org.apache.commons.lang.ObjectUtils;
 import org.apache.commons.lang.StringUtils;
 import org.apache.commons.lang.math.NumberUtils;
+import org.openhab.binding.homematic.internal.misc.MiscUtils;
 import org.openhab.binding.homematic.internal.model.HmDatapoint;
 import org.openhab.binding.homematic.internal.model.HmParamsetType;
 import org.openhab.binding.homematic.internal.model.HmValueType;
@@ -110,7 +111,8 @@ public abstract class CommonRpcParser<M, R> implements RpcParser<M, R> {
      * Returns the address of a device, replacing group address identifier.
      */
     protected String getAddress(Object object) {
-        return StringUtils.replaceOnce(toString(object), "*", "T-");
+        String address = StringUtils.trimToNull(StringUtils.replaceOnce(toString(object), "*", "T-"));
+        return MiscUtils.validateCharacters(address, "Address", "_");
     }
 
     /**

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/DeleteDevicesParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/DeleteDevicesParser.java
@@ -26,7 +26,7 @@ public class DeleteDevicesParser extends CommonRpcParser<Object[], List<String>>
         if (message != null && message.length > 1) {
             Object[] data = (Object[]) message[1];
             for (int i = 0; i < message.length; i++) {
-                String address = getAddress(data[i]);
+                String address = getSanitizedAddress(data[i]);
                 boolean isDevice = !StringUtils.contains(address, ":")
                         && !StringUtils.startsWithIgnoreCase(address, "BidCos");
                 if (isDevice) {

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/DeleteDevicesParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/DeleteDevicesParser.java
@@ -26,7 +26,7 @@ public class DeleteDevicesParser extends CommonRpcParser<Object[], List<String>>
         if (message != null && message.length > 1) {
             Object[] data = (Object[]) message[1];
             for (int i = 0; i < message.length; i++) {
-                String address = toString(data[i]);
+                String address = getAddress(data[i]);
                 boolean isDevice = !StringUtils.contains(address, ":")
                         && !StringUtils.startsWithIgnoreCase(address, "BidCos");
                 if (isDevice) {

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/EventParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/EventParser.java
@@ -35,7 +35,7 @@ public class EventParser extends CommonRpcParser<Object[], HmDatapointInfo> {
             channel = HmChannel.CHANNEL_NUMBER_VARIABLE;
         } else {
             String[] configParts = StringUtils.trimToEmpty(addressWithChannel).split(":");
-            address = getAddress(configParts[0]);
+            address = getSanitizedAddress(configParts[0]);
             if (configParts.length > 1) {
                 channel = NumberUtils.createInteger(configParts[1]);
             }

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/HomegearLoadDeviceNamesParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/HomegearLoadDeviceNamesParser.java
@@ -41,7 +41,7 @@ public class HomegearLoadDeviceNamesParser extends CommonRpcParser<Object[], Voi
             String id = toString(data.get("ID"));
             String name = toString(data.get("NAME"));
 
-            HmDevice device = devicesById.get(getAddress(id));
+            HmDevice device = devicesById.get(getSanitizedAddress(id));
             if (device != null) {
                 device.setName(name);
             }

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/ListBidcosInterfacesParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/ListBidcosInterfacesParser.java
@@ -33,7 +33,7 @@ public class ListBidcosInterfacesParser extends CommonRpcParser<Object[], ListBi
                 if (isDefault) {
                     type = toString(mapMessage.get("TYPE"));
                     firmware = toString(mapMessage.get("FIRMWARE_VERSION"));
-                    gatewayAddress = getAddress(mapMessage.get("ADDRESS"));
+                    gatewayAddress = getSanitizedAddress(mapMessage.get("ADDRESS"));
                 }
             }
         }

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/ListBidcosInterfacesParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/ListBidcosInterfacesParser.java
@@ -33,7 +33,7 @@ public class ListBidcosInterfacesParser extends CommonRpcParser<Object[], ListBi
                 if (isDefault) {
                     type = toString(mapMessage.get("TYPE"));
                     firmware = toString(mapMessage.get("FIRMWARE_VERSION"));
-                    gatewayAddress = toString(mapMessage.get("ADDRESS"));
+                    gatewayAddress = getAddress(mapMessage.get("ADDRESS"));
                 }
             }
         }

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/ListDevicesParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/ListDevicesParser.java
@@ -45,7 +45,7 @@ public class ListDevicesParser extends CommonRpcParser<Object[], Collection<HmDe
             boolean isDevice = !StringUtils.contains(toString(data.get("ADDRESS")), ":");
 
             if (isDevice) {
-                String address = getAddress(data.get("ADDRESS"));
+                String address = getSanitizedAddress(data.get("ADDRESS"));
                 String type = MiscUtils.validateCharacters(toString(data.get("TYPE")), "Device type", "-");
                 String id = toString(data.get("ID"));
                 String firmware = toString(data.get("FIRMWARE"));
@@ -54,7 +54,7 @@ public class ListDevicesParser extends CommonRpcParser<Object[], Collection<HmDe
                         new HmDevice(address, hmInterface, type, config.getGatewayInfo().getId(), id, firmware));
             } else {
                 // channel
-                String deviceAddress = getAddress(data.get("PARENT"));
+                String deviceAddress = getSanitizedAddress(data.get("PARENT"));
                 HmDevice device = devices.get(deviceAddress);
 
                 String type = toString(data.get("TYPE"));

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/NewDevicesParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/NewDevicesParser.java
@@ -34,7 +34,7 @@ public class NewDevicesParser extends CommonRpcParser<Object[], List<String>> {
                 boolean isDevice = !StringUtils.contains(address, ":")
                         && !StringUtils.startsWithIgnoreCase(address, "BidCos");
                 if (isDevice) {
-                    adresses.add(address);
+                    adresses.add(getAddress(address));
                 }
             }
         }

--- a/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/NewDevicesParser.java
+++ b/addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/parser/NewDevicesParser.java
@@ -34,7 +34,7 @@ public class NewDevicesParser extends CommonRpcParser<Object[], List<String>> {
                 boolean isDevice = !StringUtils.contains(address, ":")
                         && !StringUtils.startsWithIgnoreCase(address, "BidCos");
                 if (isDevice) {
-                    adresses.add(getAddress(address));
+                    adresses.add(getSanitizedAddress(address));
                 }
             }
         }

